rclpy/rclpy/action/server.py

```diff
@@ -0,0 +1,534 @@
+
+        # Import the typesupport for the action module if not already done
+        check_for_type_support(action_type)
+        self._node = node
+        self._action_type = action_type
+        self._handle = _rclpy_action.rclpy_action_create_server(
+            node.handle,
+            node.get_clock().handle,
+            action_type,
+            action_name,
+            goal_service_qos_profile.get_c_qos_profile(),
+            result_service_qos_profile.get_c_qos_profile(),
+            cancel_service_qos_profile.get_c_qos_profile(),
+            feedback_pub_qos_profile.get_c_qos_profile(),
+            status_pub_qos_profile.get_c_qos_profile(),
+            result_timeout,
+        )
+
+        # key: UUID in bytes, value: GoalHandle
+        self._goal_handles = {}
+
+        callback_group.add_entity(self)
+        self._node.add_waitable(self)
+
+    async def _execute_goal_request(self, request_header_and_message):
+        request_header, goal_request = request_header_and_message
+        goal_uuid = goal_request.action_goal_id
+        goal_info = GoalInfo()
+        goal_info.goal_id = goal_uuid
+        goal_info.stamp = self._node.get_clock().now().to_msg()
+
+        self._node.get_logger().debug('New goal request with ID: {0}'.format(goal_uuid.uuid))
+
+        # Check if goal ID is already being tracked by this action server
+        with self._lock:
+            goal_id_exists = _rclpy_action.rclpy_action_server_goal_exists(self._handle, goal_info)
+
+        accepted = False
+        if not goal_id_exists:
+            # Call user goal callback
+            response = await await_or_execute(self._goal_callback, goal_request)
+            accepted = ((GoalResponse.ACCEPT_AND_EXECUTE == response) or
+                        (GoalResponse.ACCEPT_AND_DEFER == response))
+
+        if accepted:
+            # Create a goal handle
+            try:
+                with self._lock:
+                    goal_handle = ServerGoalHandle(self, goal_info, goal_request)
+            except RuntimeError as e:
+                self._node.get_logger().error(
+                    'Failed to accept new goal with ID {0}: {1}'.format(goal_uuid.uuid, e))
+                accepted = False
+            else:
+                self._goal_handles[bytes(goal_uuid.uuid)] = goal_handle
+
+        # Send response
+        response_msg = self._action_type.GoalRequestService.Response()
+        response_msg.accepted = accepted
+        response_msg.stamp = self._node.get_clock().now().to_msg()
+        _rclpy_action.rclpy_action_send_goal_response(
+            self._handle,
+            request_header,
+            response_msg,
+        )
+
+        if not accepted:
+            self._node.get_logger().debug('New goal rejected: {0}'.format(goal_uuid.uuid))
+            return
+
+        self._node.get_logger().debug('New goal accepted: {0}'.format(goal_uuid.uuid))
+
+        goal_handle._update_state(GoalEvent.EXECUTE)
```

Comment:
This looks like it is setting the goal handle to `EXECUTE` even if the user chose `ACCEPT_AND_DEFER`. In `rclcpp` instead of an `execute_callback` there is `handle_accepted`. If the user says `ACCEPT_AND_DEFER` then it arrives at that callback in state `ACCEPTED`.

If an `execute_callback` is provided then it should be separate from `handle_accepted`. Maybe add the `execute_callback` as a task to the executor if `handle_accepted` returns and the goal handle is not at a terminal state?

Reply:
I... deferred implementing the defer logic. I meant to make a note of it.

Originally I was thinking that we could provide an API via the `GoalHandle` to execute a deferred goal. E.g. `def execute(callback=None)` where the callback will optionally override a previous established callback. This function will schedule the callback (if there is one) with the executor.

I like the idea of the `execute_callback` as a way to let the action server handle scheduling execution with the executor. I think this puts less burden on the user; not having to manage threads. Do you think this is something `rclcpp_action` is missing?

Reflecting on the `rclcpp` action server, I think it makes sense to take a `handle_accepted` callback in `rclpy`'s aciton server as well. The `handle_accepted` makes a nice venue for providing the user with a reference to the goal handle.

Reply:
> I like the idea of the `execute_callback` as a way to let the action server handle scheduling execution with the executor. I think this puts less burden on the user; not having to manage threads. Do you think this is something `rclcpp_action` is missing?

I agree an `execute_callback` is convenient. I've assumed in `rclcpp_action` an execute callback would be provided by a class like ROS 1's `SimpleActionServer`. It should only be invoked after a goal transitions to `EXECUTING` though. Right now I don't see a way to make a python action server queue goals instead of executing them all in parallel.

Looking back, I'm not sure `ACCEPT_AND_EXECUTE` and `ACCEPT_AND_DEFER` are necessary. The same thing could be accomplished by whether or not `handle_goal` changed the state to `EXECUTING` .

Reply:
> Looking back, I'm not sure `ACCEPT_AND_EXECUTE` and `ACCEPT_AND_DEFER` are necessary.

Agreed.

I've made an update in 9a15695 that adds a `handle_accepted_callback` in which the user can optionally call `goal_handle.execute()`, or hold onto the goal handle to execute later (ie. defer).

Reply:
> Right now I don't see a way to make a python action server queue goals instead of executing them all in parallel

With the updated examples, I believe the equivalent of a `SimpleActionServer` could be implemented in Python by adapting the ["single goal" example](https://github.com/ros2/examples/blob/0ea84a050d5025fde8ab65000dc3579937ee2cf7/rclpy/actions/minimal_action_server/examples_rclpy_minimal_action_server/server_single_goal.py). Instead of aborting newly received goals, it could queue them instead.